feat(web-analytics): Display data relevant to currently selected conversion goal

rafaeelaudibert opened the pull request:
Customers can select a conversion goal at the top of the web analytics panel. We were, however, not respecting that conversion goal when rendering most of the panels in the web analytics screen.

Now, we properly respect that in the backend + hide the non-relevant parts of the UI, i.e. the ones that can't have their conversion rate computed.

## Screenshots

![image](https://github.com/user-attachments/assets/fdb48908-2b23-45a2-af10-b2615ca9fd60)


## Tests

UI locally + new unit tests

rafaeelaudibert committed:
WIP adding support to our queries to display only results related to the currently selected `conversionGoal`

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -1082,6 +1087,7 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
                               },
                           }
                         : null,
+                    // TODO: What to do with this one if conversion goal is set? Hide it? Disable it?
```
Any ideas @robbie-c? I believe we should simply hide the parts of the UI that aren't "filtered" by conversion goals when this happens.

robbie-c replied:
Yeah good Q.

One idea could be that it works like list of funnels. E.g. if you have Action G as the goal, and Actions A B C as other actions, it could shows

Action A - some conversion stats from A -> G
Action B - some conversion stats from B -> G
Action C - some conversion stats from C -> G

though maybe it should be the other way around? e.g. G -> A

I think it's also fine to just hide this tile when a conversion goal is set :)

rafaeelaudibert replied:
Decision for now: just keep it hidden, but this would be a cool visualization :)

rafaeelaudibert commented on posthog/hogql_queries/web_analytics/stats_table.py:
Extremely WIP file

rafaeelaudibert committed:
Stop displaying unrelated-to-conversion-goal sections

rafaeelaudibert committed:
Add `conversion` support to session and entry bounce queries

rafaeelaudibert committed:
Add conversion rate column

rafaeelaudibert committed:
Add conversion tests

rafaeelaudibert commented on posthog/hogql_queries/web_analytics/test/test_web_stats_table.py:
```diff
@@ -1255,3 +1278,216 @@ def test_timezone_filter_with_empty_timezone(self):
+            "2023-12-01", "2023-12-03", breakdown_by=WebStatsBreakdown.PAGE, custom_event="custom_event"
+        )
+
+        # TODO: It's returning empty, I don't know why
+        assert [["https://www.example.com/foo", (1, 0), (0, 0), (0, 0), (0, None)]] == response.results
+        assert [
+            "context.columns.breakdown_value",
+            "context.columns.visitors",
+            "context.columns.total_conversions",
+            "context.columns.unique_conversions",
+            "context.columns.conversion_rate",
+        ] == response.columns
```
@robbie-c Ok, I know why. We're filtering events by `$pageView` in all of our queries. What should happen in this case when we're trying to handle conversion? Should this simply not work? Or should we drop the `$pageView` filter?

robbie-c replied:
In the overview query, if there's a conversion goal set, I change the filter from event=pathname to (event=pathname OR {conversion_goal_expr})

rafaeelaudibert commented:
@robbie-c I've got this working for everything besides "Page" (`Path` in the UI for some reason) queries
![image](https://github.com/user-attachments/assets/71567d16-a599-4592-9cd9-d50b286ac19a)

Their queries are very different and will require me more work to get them to work. 2 questions:
- does it make sense to ship this PR before I can get them to work?
- does it even make sense to compute conversion rate for that?

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -1267,7 +1274,7 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
                           }
                         : null,
-                    featureFlags[FEATURE_FLAGS.WEB_ANALYTICS_REPLAY]
+                    !conversionGoal && featureFlags[FEATURE_FLAGS.WEB_ANALYTICS_REPLAY]
```
TODO: Might check with Paul if there's an easy way to display recordings associated with a conversion

rafaeelaudibert committed:
Render session replays relevant to the chosen conversion goal

rafaeelaudibert committed:
Move conversation rate computation to HogQL

Much more performant, and much simpler code

rafaeelaudibert committed:
Fix all conversion-related test

rafaeelaudibert committed:
Span Replay the whole width if we're not rendering error tracking

rafaeelaudibert committed:
Return conversion for `PAGE` queries

Simply use the main query, scroll bounce/bounce doesn't make sense

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -1418,12 +1426,22 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
+                shouldFilterTestAccounts,
+                conversionGoal
             ): RecordingUniversalFilters => {
+                const filters: UniversalFiltersGroupValue[] = [...webAnalyticsFilters]
+                if (conversionGoal) {
+                    const actionId = (conversionGoal as ActionConversionGoal).actionId
+                    const customEventName = (conversionGoal as CustomEventConversionGoal).customEventName
+                    const id = actionId || customEventName
+
+                    filters.push({ id, name: String(id), type: customEventName ? 'events' : 'actions' })
+                }
```
Is there an easier way to do this, maybe?

robbie-c replied:
Yes definitely, you can do 

`if "actionId" in conversionGoal` or `if "customEventName" in conversionGoal` and TypeScript will do [narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html) to give you the correct types inside that branch of the if statement

rafaeelaudibert commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -1418,12 +1426,22 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
+                shouldFilterTestAccounts,
+                conversionGoal
             ): RecordingUniversalFilters => {
+                const filters: UniversalFiltersGroupValue[] = [...webAnalyticsFilters]
+                if (conversionGoal) {
+                    const actionId = (conversionGoal as ActionConversionGoal).actionId
+                    const customEventName = (conversionGoal as CustomEventConversionGoal).customEventName
+                    const id = actionId || customEventName
+
+                    filters.push({ id, name: String(id), type: customEventName ? 'events' : 'actions' })
+                }
```
@pauldambra Pinging you to learn if querying replay sessions this way makes sense or if I won't get the results I'm expecting. This `filters` entry is parsed into either the `actions` or `events` parameter that's sent to the backend.

I know `name` should be the name of the action rather than its id, but we don't have access to that very easily in this part of the app. Is it ok if `name` is not really the name? I assume it's fine and this is used for displaying purposes only - which we don't need.

pauldambra replied:
I don't actually know 🙈 

```
action = Action.objects.create(
            team_id=team_id,
            name=name,
            steps_json=[
                {
                    "event": name,
                    "properties": properties,
                }
            ],
        )
```

when setting up backend tests for filtering we're creating actions like this which suggests we care about the name and not the id 😬

rafaeelaudibert replied:
Welp, that's not good. I'll dive into that part of the code to learn if this will work or not, thx!

rafaeelaudibert committed:
Fix mypy complaints

rafaeelaudibert committed:
Simplify TS checks

rafaeelaudibert committed:
Protect conversion goal behind feature flag

robbie-c commented:
Can we add some text to the paths popover to explain what counts as a conversion? i.e. does the conversion need to happen on that path to count? Or is it all the conversions where the user visited that path at least once during the session? Or where they had a pageview visiting that path before the conversion.

rafaeelaudibert commented:
> Can we add some text to the paths popover to explain what counts as a conversion in the paths insight? i.e. does the conversion need to happen on that path to count? Or is it all the conversions where the user visited that path at least once during the session? Or where they had a pageview visiting that path before the conversion.

@robbie-c Of course :). It should be the first one: conversion happened on that page. My understanding of the code is still low, so that's why I want you to check the code to see if that's really what it's doing.

I admit that's slightly useless, though, so we might wanna change that in the future.

rafaeelaudibert committed:
docs: Improve documentation when conversion goal is selected

robbie-c commented on posthog/hogql_queries/web_analytics/stats_table.py:
```diff
@@ -27,6 +32,8 @@
 BREAKDOWN_NULL_DISPLAY = "(none)"
 
 
+# TODO: Extend `conversion_goal` support to queries besides `to_main_query`
```
I think this isn't a to-do any more, we only need to support the main query.

The other queries are to show bounce rate, scroll depth, etc, but we don't need to do that with a conversion goal active

robbie-c commented on posthog/hogql_queries/web_analytics/stats_table.py:
```diff
@@ -27,6 +32,8 @@
 BREAKDOWN_NULL_DISPLAY = "(none)"
 
 
+# TODO: Extend `conversion_goal` support to queries besides `to_main_query`
```
I think this isn't a to-do any more, we only need to support the main query.

The other queries are to show bounce rate, scroll depth, etc, but we don't need to do that with a conversion goal active

rafaeelaudibert replied:
That's correct, forgot to remove the comments, thank you

robbie-c commented on posthog/hogql_queries/web_analytics/stats_table.py:
```diff
@@ -438,6 +343,124 @@ def to_path_bounce_query(self) -> ast.SelectQuery:
+    any(person_id) AS filtered_person_id,
+    count() AS filtered_pageview_count,
+    {breakdown_value} AS breakdown_value,
+    session.session_id AS session_id,
```
A few notes on this:
* include_session_properties was previously passed in so that we could do a performance optimisation, where if there were no session properties, we could avoid joining on the sessions table altogether. This meant we had to group by `events.$session_id` rather than `session.session_id`.
* This optimisation meant that queries with an event property breakdown (e.g. the geoip and device ones) would be a little bit faster
* With how you're comparing to previous periods using the session start timestamp, we're always using the session table, so it's no longer a useful distinction, that performance optimisation can never happen any more
* Hypothetically, we could bring back the performance optimisation, by using the event timestamp in `period_aggregate` rather than the session start timestamp, for queries that don't touch the sessions table
* IMO we don't need to bring back that performance optimization in this PR. First, let's see what the performance is like for loading all queries on the page in one go

rafaeelaudibert replied:
Done, I've removed the optimization from most queries. The ones that run when no conversion goal is set have been kept.

robbie-c commented on posthog/hogql_queries/web_analytics/stats_table.py:
```diff
@@ -438,6 +343,124 @@ def to_path_bounce_query(self) -> ast.SelectQuery:
+
+        assert isinstance(query, ast.SelectQuery)
+
+        if include_session_properties:
```
Related to the comment above

robbie-c commented on posthog/hogql_queries/web_analytics/stats_table.py:
```diff
@@ -438,6 +343,124 @@ def to_path_bounce_query(self) -> ast.SelectQuery:
+    def _previous_period_aggregate(self, function_name, column_name):
+        return self.period_aggregate(function_name, column_name, self._date_from_previous_period(), self._date_from())
+
+    # Reproduction from `web_analytics/web_overview.py`
```
Could these functions move to the parent class to avoid duplication?

rafaeelaudibert replied:
I'll need to create a new parent class because `conversion_goal_expr` doesn't exist on the base query. I'll see how that goes, I agree it might be better than reproducing them here.

Deep down I'm ok with this because we'll eventually have a single query powering web analytics, and then there won't be any duplication anymore

rafaeelaudibert replied:
It was actually easy to extract this, done ✅

rafaeelaudibert committed:
Remove outdated TODO comments

rafaeelaudibert committed:
perf: Remove session optimization

See https://github.com/PostHog/posthog/pull/26586#discussion_r1875855669

rafaeelaudibert committed:
chore: Extract shared functions to parent

Now that all of the queries share a `conversionGoal` parameter, Python is happy with this!

robbie-c reviewed (approved):
I looked at the previous version of this in a lot of detail, so whist I haven't looked at the latest refactor in as much detail, it broadly looks good and I'm happy with the amount of test coverage, so 🚢
